extmod/network_ninaw10.c

```diff
@@ -85,25 +134,42 @@ STATIC mp_obj_t network_ninaw10_active(size_t n_args, const mp_obj_t *args) {
                     MP_ERROR_TEXT("Failed to initialize Nina-W10 module, error: %d\n"), error);
             }
             // check firmware version
-            uint8_t fw_ver[NINA_FW_VER_LEN];
-            if (nina_fw_version(fw_ver) != 0) {
+            uint8_t semver[NINA_FW_VER_LEN];
+            if (nina_fw_version(semver) != 0) {
                 nina_deinit();
                 mp_raise_msg_varg(&mp_type_OSError,
                     MP_ERROR_TEXT("Failed to read firmware version, error: %d\n"), error);
             }
-            // Check fw version matches the driver.
-            if ((fw_ver[NINA_FW_VER_MAJOR_OFFS] - 48) != NINA_FW_VER_MAJOR ||
-                (fw_ver[NINA_FW_VER_MINOR_OFFS] - 48) != NINA_FW_VER_MINOR ||
-                (fw_ver[NINA_FW_VER_PATCH_OFFS] - 48) != NINA_FW_VER_PATCH) {
-                mp_printf(&mp_plat_print,
-                    "Warning: firmware version mismatch, expected %d.%d.%d found: %d.%d.%d\n",
-                    NINA_FW_VER_MAJOR, NINA_FW_VER_MINOR, NINA_FW_VER_PATCH,
-                    fw_ver[NINA_FW_VER_MAJOR_OFFS] - 48,
-                    fw_ver[NINA_FW_VER_MINOR_OFFS] - 48,
-                    fw_ver[NINA_FW_VER_PATCH_OFFS] - 48);
+            // Check the minimum supported firmware version.
+            uint32_t fwmin = (NINA_FW_VER_MIN_MAJOR * 100) +
+                (NINA_FW_VER_MIN_MINOR * 10) +
+                (NINA_FW_VER_MIN_PATCH * 1);
+
+            uint32_t fwver = (semver[NINA_FW_VER_MAJOR_OFFS] - 48) * 100 +
+                (semver[NINA_FW_VER_MINOR_OFFS] - 48) * 10 +
+                (semver[NINA_FW_VER_PATCH_OFFS] - 48) * 1;
+
+            if (fwver < fwmin) {
+                mp_raise_msg_varg(&mp_type_OSError,
+                    MP_ERROR_TEXT("Firmware version mismatch. Minimum supported firmware is v%d.%d.%d found v%d.%d.%d\n"),
+                    NINA_FW_VER_MIN_MAJOR, NINA_FW_VER_MIN_MINOR, NINA_FW_VER_MIN_PATCH, semver[NINA_FW_VER_MAJOR_OFFS] - 48,
+                    semver[NINA_FW_VER_MINOR_OFFS] - 48, semver[NINA_FW_VER_PATCH_OFFS] - 48);
             }
+            mp_sched_lock();
```

Comment:
Why are these lock/unlock calls here, what are you trying to protect against?  The scheduler only runs if the VM runs (or mp_hal_delay), which is not the case if you just set some variables (unless it's running with threading... which maybe what you're protecting against?).

But for this particular case, if `machine_timer_type.make_new` raises an exception then `mp_sched_unlock` won't be called.  So I think you don't need this particular lock/unlock sequence here.

Reply:
If the interface is already active, and `active()` is called again, the timer will be running. Isn't it possible that timer schedules the callback with `mp_sched_schedule_node` and the poll list is accessed ? Anyway I was trying to be extra careful, but will remove them.

Reply:
Yes that makes sense.